src/ucp/core/ucp_mm.c

```diff
@@ -400,99 +425,152 @@ ucp_memh_get_slow(ucp_context_h context, void *address, size_t length,
+              ucp_md_map_t reg_md_map, unsigned uct_flags)
 {
-    ucp_md_map_t reg_md_map       = context->reg_md_map[memory_type];
-    ucp_md_index_t alloc_md_index = UCP_NULL_RESOURCE;
-    ucp_mem_h memh                = NULL;
+    void *address = ucp_memh_address(memh);
+    size_t length = ucp_memh_length(memh);
     ucp_md_index_t md_index;
-    uct_allocated_memory_t mem;
     ucs_status_t status;
 
-    status = ucp_mem_do_alloc(context, address, length, uct_flags, memory_type,
-                              alloc_name, &mem);
-    if (status != UCS_OK) {
-        return status;
-    }
+    if (context->rcache == NULL) {
+        status = ucp_memh_register(context, memh, reg_md_map, address, length,
+                                   uct_flags);
+        if (status != UCS_OK) {
+            return status;
+        }
 
-    if (mem.method == UCT_ALLOC_METHOD_MD) {
-        for (md_index = 0; md_index < context->num_mds; md_index++) {
-            if (mem.md == context->tl_mds[md_index].md) {
-                alloc_md_index = md_index;
-                reg_md_map    &= ~UCS_BIT(md_index);
-                break;
-            }
+        memh->parent = memh;
+    } else {
+        status = ucp_memh_get(context, address, length, memh->mem_type,
+                              reg_md_map, uct_flags, &memh->parent);
+        if (status != UCS_OK) {
+            return status;
         }
 
-        ucs_assert(alloc_md_index != UCP_NULL_RESOURCE);
+        ucs_for_each_bit(md_index, reg_md_map) {
+            memh->uct[md_index] = memh->parent->uct[md_index];
+            memh->md_map       |= UCS_BIT(md_index);
+        }
     }
 
-    status = ucp_memh_get_slow(context, mem.address, mem.length,
-                               mem.mem_type, reg_md_map, uct_flags, &memh);
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_memh_alloc(ucp_context_h context, size_t length,
+               ucs_memory_type_t mem_type, ucp_mem_h *memh_p)
+{
+    uct_allocated_memory_t mem;
+    ucs_status_t status;
+    ucp_mem_h memh;
+
+    status = ucp_memh_create_no_rcache(context, NULL, length, mem_type,
+                                       &memh);
     if (status != UCS_OK) {
-        goto err;
+        goto out;
     }
 
-    memh->alloc_method = mem.method;
-    if (mem.method == UCT_ALLOC_METHOD_MD) {
-        memh->alloc_md_index      = alloc_md_index;
-        memh->uct[alloc_md_index] = mem.memh;
-        memh->md_map             |= UCS_BIT(alloc_md_index);
-        ucs_trace("allocated address %p length %zu on md[%d]=%s %p",
-                mem.address, mem.length, alloc_md_index,
-                context->tl_mds[alloc_md_index].rsc.md_name,
-                memh->uct[alloc_md_index]);
+    status = ucp_mem_do_alloc(context, NULL, length, UCT_MD_MEM_ACCESS_ALL,
+                              mem_type, &mem);
+    if (status != UCS_OK) {
+        goto err_free_memh;
+    }
+
+    ucp_memh_init_from_mem(context, memh, &mem);
```

Comment:
seems like ucp_memh_init_from_mem overrides memh fields that were set by ucp_memh_create_no_rcache ?
maybe we can first call ucp_mem_do_alloc and only then initialize the memh once with the right fields?